File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/levels/traps/GrippingTrap.java

package com.watabou.pixeldungeon.levels.traps;

import com.nyrds.pixeldungeon.levels.objects.ITrigger;
import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.buffs.Bleeding;
import com.watabou.pixeldungeon.actors.buffs.Buff;
import com.watabou.pixeldungeon.actors.buffs.Cripple;
import com.watabou.pixeldungeon.effects.Wound;

import org.jetbrains.annotations.Nullable;

public class GrippingTrap implements ITrigger {
	
	public static void trigger( int pos, @Nullable Char c ) {
		
		if (c != null) {
			int damage = Math.max( 0,  (Dungeon.depth + 3) - c.defenceRoll(c)/2);
			Buff.affect( c, Bleeding.class ).level( damage );
			Buff.prolong( c, Cripple.class, Cripple.DURATION );
			Wound.hit( c );
		} else {
			Wound.hit( pos );
		}
	}

	@Override
	public void doTrigger(int cell, Char ch) {
		trigger(cell,ch);
	}
}

File: scripts/traps/GrippingTrap.lua
local RPD = require "scripts/lib/commonClasses"
local trap = require "scripts/lib/trap"

-- java GrippingTrap: crushes the foot - bleeding scaled by depth vs defence,
-- cripple, wound visual. Wound-only if triggered with nobody on the cell.
return trap.init(
    function (cell, char, data)
        if char ~= nil then
            local damage = math.max(0, (RPD.Dungeon.depth + 3) - math.floor(char:defenceRoll(char) / 2))
            local bleeding = RPD.Buffs.Buff:affect(char, "Bleeding")
            bleeding:level(damage)
            RPD.Buffs.Buff:prolong(char, "Cripple", 10)
            RPD.Sfx.Wound:hit(char)
        else
            RPD.Sfx.Wound:hit(cell)
        end
    end
)
